add user-submitted clips page

trimbe opened the pull request:
## Describe your changes

#411

Adds clips page for users to submit and vote on clips from the alveus or maya channels. Clips can be sorted by top/top 7 days/top 30 days/new. Clips require approval before they can be viewed publicly and an admin panel is added to facilitate that. A daily cron is added to verify validity of submitted clips and remove them if the clip has been deleted. 

Some thoughts: 

I'm not at all attached to the design, I primarily was going for something functional and not ugly. Very open to suggestions to make it look better.

Voting currently requires the user to be logged in, but that might not be great for engagement. Possibly allow logged out voting, but add a unique constraint on clip slug/IP address for the ClipVotes to mitigate potential manipulation?

I like the idea of a "hot" sort, but it seems this would be less useful unless there's some threshold of voting traffic and I'm not sure how much traffic the website can be expected to receive. 

I like many of the other ideas in #411 like ambassador tagging, leaderboard/points of some sort, and deep share links, but wanted to submit this base version first. Happy to add on later if these other features are desirable. 

## Notes for testing your change

Deployed here: https://alveusgg-website.vercel.app/clips (don't think Vercel previews can handle the db changes?)

pjeweb commented:
@trimbe this looks amazing. i will look into the code as soon as i can. still figuring out how to apply the schema changes for the PR to a new db branch, so no preview deployment right now.

pjeweb commented on apps/website/prisma/schema.prisma:
```diff
@@ -454,3 +456,31 @@ model VirtualTicket {
+
+model Clip {
+  id            String @id @default(cuid())
+  clipSlug      String @unique
```
Is there a specific reason we have use the clip prefix here?

pjeweb commented on apps/website/prisma/schema.prisma:
```diff
@@ -454,3 +456,31 @@ model VirtualTicket {
+  title         String
+  thumbnail     String
+  createdAt     DateTime @default(now())
+  clipCreator   String?
```
dito

pjeweb commented on apps/website/prisma/schema.prisma:
```diff
@@ -454,3 +456,31 @@ model VirtualTicket {
+  thumbnail     String
+  createdAt     DateTime @default(now())
+  clipCreator   String?
+  submitted     DateTime @default(now())
```
```suggestion
  submittedAt     DateTime @default(now())
```
to be consistent

pjeweb commented on apps/website/prisma/schema.prisma:
```diff
@@ -454,3 +456,31 @@ model VirtualTicket {
+  clipSlug      String @unique
+  userId        String?
+  title         String
+  thumbnail     String
```
Is this an id or url? Can we call it `thumbnailUrl` or `thumbnailId`?

pjeweb commented on apps/website/src/server/trpc/router/clips.ts:
```diff
@@ -0,0 +1,142 @@
+      // https://m.twitch.tv/channel/clip/slug?query
+
+      const clipRegex =
+        /^https?:.+(?:clips\.twitch\.tv|(?:m\.)?twitch\.tv\/(?:clip|.+\/clip))\/?([\w\d-]+)/;
```
```suggestion
        /^https?:.+(?:clips\.twitch\.tv|(?:m\.)?twitch\.tv\/(?:clip|.+\/clip))\/?([\w-]+)/;
```

`\w` already includes `[0-9]`

pjeweb commented on apps/website/src/server/trpc/router/clips.ts:
```diff
@@ -0,0 +1,142 @@
+        });
+      }
+
+      const channelAllowlist = ["alveussanctuary", "maya"];
```
Probably wanna move this to an env variable or config file

pjeweb commented on apps/website/src/server/trpc/router/clips.ts:
```diff
@@ -0,0 +1,142 @@
+      ) {
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Clip not from AlveusSanctuary or Maya",
```
Would keep the message generic or based on the allow list?

pjeweb reviewed (commented):
I did not have time to check out the code 100%, but I'm overall pretty impressed! Would appreciate feedback on the code base and how it was to work with too :)

Some initial feedback/ideas:
- Appreciate the MVP. I also think, its best we start small.
- Design overall is pretty good and clean. I only have small things I'd wanna change.
- I did not have that in #414 and I am sorry for that, but I was thinking it might be better to add _all_ created clips to the site. At least the ones from the Alveus channel. I think one problem with the smashscene is, that people usually only submit their own clips and not many do since its an additional step.
- In the same vein, I think it might be better to accept all clips without approval by default and only allow to remove approval.

trimbe commented:
> I did not have that in https://github.com/alveusgg/alveusgg/issues/411 and I am sorry for that, but I was thinking it might be better to add all created clips to the site. At least the ones from the Alveus channel. I think one problem with the smashscene is, that people usually only submit their own clips and not many do since its an additional step.

Yeah this sounds sensible, doesn't look like there's any events in EventSub so I'll add a cron for this.

> In the same vein, I think it might be better to accept all clips without approval by default and only allow to remove approval.

Alright, sounds good, I just wanted to err on the side of caution for moderation purposes but it's probably fine.

> Would appreciate feedback on the code base and how it was to work with too :)

This is my first time using React/next/prisma in a serious project, but from that perspective, I didn't really have any issues finding my way around or understanding the flow.

trimbe commented:
@pjeweb So I've written a cronjob to add all clips but now that I'm looking at some of the clips I'm not sure if it's the best idea. It appears that clips created on mobile are saved as soon as the clip button is pressed and given a default title of the stream's title at the time. A pretty significant proportion of clips are these default clips, and I'm not sure how often anything interesting happens in them (misclicks, or people just using them to rewind?). I feel like they just add clutter if you're trying to look through the new clips.

Filtering by views when added may be a good idea. A floor of 10 views reduces the total clips on AlveusSanctuary from 13400 to 7600 and I think that still catches the interesting stuff if people organically share with friends or chat. I'd add some buffer time when adding to allow time for the clips to accumulate views.

Another consideration is overlapping clips, where several users clip the same moment. I'm not sure this is really possible to avoid intelligently, but it also might be mitigated to some extent by the views requirement.

trimbe commented:
@pjeweb Clip view distribution: 

| Views     | Number |
|-----------|--------|
| 1-5       | 4,247  |
| 6-10      | 1,858  |
| 11-50     | 4,350 |
| 51-100    | 1,330  |
| 101-500   | 1,597  |
| 501-1000  | 223    |
| 1001-5000 | 130    |
| 5000+     | 37     |

I do like the chatbot idea but I worry that will exacerbate the overlapping clip issue. I think with a sane view requirement, we probably accomplish pretty much the same thing anyway. 50+ seems like it may be alright, and I'll stagger it by 10 minutes or so to allow for accumulation of views. And allow non-logged-in submissions.

For the initial import I can add a postinstall script, and increase the threshold to 500+ maybe.

trimbe commented:
Addressed your reviews and made changes as per my previous comment. Not fully confident on the 30m/50 view numbers, might be a bit high for slow days. None of the clips from today would have been added, for example, but maybe nothing warranted it.

pjeweb commented:
@trimbe thank you! might take me a few days to review though. sorry about that

pjeweb commented:
@trimbe would you mind me pushing a merge commit with the updates to the show and tell pages to your branch?

trimbe commented:
@pjeweb Nope, go ahead.

pjeweb commented on apps/website/src/pages/clips/submit-clip.tsx:
```diff
@@ -0,0 +1,128 @@
+import Image from "next/image";
+import type { FormEvent } from "react";
+import { useState } from "react";
+import { useSession } from "next-auth/react";
```
`useSession` is not being used

pjeweb commented on apps/website/src/pages/clips/index.tsx:
```diff
@@ -0,0 +1,330 @@
+
+  return (
+    <div>
+      <div
+        ref={lastItemRef}
+        className="group cursor-pointer"
+        onClick={() => clipClicked(clip)}
+      >
```
clickable elements should be actually `<button>` elements (or `<a>`)

pjeweb commented on apps/website/src/pages/clips/index.tsx:
```diff
@@ -0,0 +1,330 @@
+      <div className="flex justify-between px-4">
+        <p className="text-left">Votes: {clip.voteCount}</p>
+        {session && (
+          <p
+            className={classes("cursor-pointer text-right", thumbsClass)}
+            onClick={() => handleVote(clip.id)}
+          >
+            <IconThumbsUp size={20} />
+          </p>
```
dito make it a `<button>`

pjeweb commented on apps/website/src/pages/clips/index.tsx:
```diff
@@ -0,0 +1,330 @@
+  clipUrl.searchParams.set("parent", window.location.hostname);
+
+  return (
+    <div
```
The modal dialog should capture keyboard focus and close when pressing the escape key and have appropriate aria attributes. The easiest would be to use the headless `Dialog` component (see https://github.com/alveusgg/alveusgg/blob/main/apps/website/src/components/shared/ModalDialog.tsx). This will require some tweaks to the event handler but it should be fairly easy to replace the divs with `Dialog`, `Dialog.Panel` and the headline with a `Dialog.Title`. Reusing the existing `ModalDialog` might also work but the button is fixed as a labeled button at the bottom of the dialog ...

pjeweb commented on apps/website/src/pages/clips/index.tsx:
```diff
@@ -0,0 +1,330 @@
+      onClickCapture={onClose}
+    >
+      <div className="relative m-auto flex max-h-[50vh] w-full flex-col rounded-lg border border-solid border-black bg-white px-2 pb-4 pt-10 sm:px-10 sm:py-10 md:max-w-[75vw] lg:max-h-[75vh] lg:max-w-[60vw] 2xl:max-w-[50vw]">
+        <span className="absolute right-0 top-0 p-2" onClick={onClose}>
+          <svg
+            onClick={onClose}
+            className="h-6 w-6 cursor-pointer text-gray-700"
+            fill="none"
+            viewBox="0 0 24 24"
+            stroke="currentColor"
+          >
+            <path
+              onClick={onClose}
+              strokeLinecap="round"
+              strokeLinejoin="round"
+              strokeWidth="2"
+              d="M6 18L18 6M6 6l12 12"
+            />
+          </svg>
+        </span>
```
dito use a `<button>`. the events should bubble up, attaching the events to the svg and path should not be necessary

pjeweb commented on apps/website/src/pages/clips/index.tsx:
```diff
@@ -0,0 +1,330 @@
+      {/* Nav background */}
+      <div className="-mt-40 hidden h-40 bg-alveus-green-900 lg:block" />
+
+      <div className="relative">
```
I think this div is not needed. This was for placing leaves i think?!

pjeweb commented on apps/website/src/pages/show-and-tell/my-posts/[postId]/index.tsx:
```diff
@@ -6,16 +6,18 @@ import { trpc } from "@/utils/trpc";
 
 import showAndTellHeader from "@/assets/show-and-tell/header.png";
 
+import { showAndTellNavItems } from "../..";
```
Note sure i like the relative path here. Maybe we can just use the full path or adding back the ShowAndTellNavigation component (basically just wrapping the `PageNavigation`) in its own file and co-locate the items there?

pjeweb commented on apps/website/src/pages/clips/submit-clip.tsx:
```diff
@@ -0,0 +1,128 @@
+  const [error, setError] = useState<string | null>(null);
+  const [submitting, setSubmitting] = useState(false);
+
+  const clipsNavigationItems = [
+    {
+      href: "/clips/",
+      label: "Clips",
+      exact: true,
+    },
+    {
+      href: "/clips/submit-clip",
+      label: "Submit Clip",
+    },
+  ];
```
Instead of repeating the items maybe create a component wrapping the `PageNavigation` in its own file and co-locate the items array there? Exporting the items from one file and reusing them in the other can also work (see comment about the relative import for show and tell).

pjeweb reviewed (changes requested):
Overall I'm very happy with your work. I wanna get @MattIPv4's eyes on the UI/UX (esp. design and text) at some point too, but I think this should make it in the site soon :)

MattIPv4 commented:
👋 Apologies @trimbe it looks like one of my recent merges has caused some conflicts here. Would you be able to get them resolved and get this branch up-to-date with main?

trimbe commented:
@MattIPv4 👍

pjeweb commented:
Sorry for the delays. Trying to find the time to do the final review.

pjeweb commented:
Just a small update: I have a bunch of small tweaks in a local branch that I need to finish. If you do not mind, I might push the changes in your PR directly and explain with comments? It will probably still take me some days to do that either way.

trimbe commented:
Yeah, that's fine, go for it.

pjeweb commented:
I am gonna close this PR for now. I think we have other people working on a standalone clips archive.

@trimbe Thank you a lot for your work and sorry we did not get it shipped.
